flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/plan/rules/physical/batch/EnforceLocalAggRuleBase.scala

```diff
@@ -0,0 +1,197 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.table.planner.plan.rules.physical.batch
+
+import org.apache.flink.table.api.TableException
+import org.apache.flink.table.planner.calcite.FlinkTypeFactory
+import org.apache.flink.table.planner.plan.`trait`.FlinkRelDistribution
+import org.apache.flink.table.planner.plan.nodes.FlinkConventions
+import org.apache.flink.table.planner.plan.nodes.physical.batch.{BatchExecExchange, BatchExecExpand, BatchExecGroupAggregateBase, BatchExecHashAggregate, BatchExecLocalHashAggregate, BatchExecLocalSortAggregate, BatchExecSortAggregate}
+import org.apache.flink.table.planner.plan.utils.{AggregateUtil, FlinkRelOptUtil}
+import org.apache.flink.table.runtime.types.LogicalTypeDataTypeConverter.fromDataTypeToLogicalType
+
+import org.apache.calcite.plan.{RelOptRule, RelOptRuleCall, RelOptRuleOperand}
+import org.apache.calcite.rel.RelNode
+import org.apache.calcite.rex.RexUtil
+import org.apache.calcite.tools.RelBuilder
+import org.apache.calcite.util.Util
+
+import scala.collection.JavaConversions._
+
+/**
+  * Planner rule that writes one phase aggregate to two phase aggregate,
+  * when the following conditions are met:
+  * 1. there is no local aggregate,
+  * 2. the aggregate has non-empty grouping and two phase aggregate strategy is enabled,
+  * 3. the input is [[BatchExecExpand]] and there is at least one expand row
+  * which the columns for grouping are all constant.
+  */
+abstract class EnforceLocalAggRuleBase(
+    operand: RelOptRuleOperand,
+    description: String)
+  extends RelOptRule(operand, description)
+  with BatchExecAggRuleBase {
+
+  protected def getBatchExecExpand(call: RelOptRuleCall): BatchExecExpand
+
+  override def matches(call: RelOptRuleCall): Boolean = {
+    val agg: BatchExecGroupAggregateBase = call.rel(0)
```

Comment:
It's better to leave the `matches` method to the child classes. It's hard to tell what will be for `call.rel(0)` is here since we don't know the pattern in base class. You can find some other ways to reuse some codes though.

Reply:
yes, it's a more clean approach